### PR TITLE
WS-2 Adding new location

### PR DESCRIPTION
**Pull request by MichalKulygin**

### CODE REVIEW

File: src/main/java/com/mkgn/weatherservice/exceptions/EmptyInputException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.mkgn.weatherservice.exceptions;
+
+public class EmptyInputException extends RuntimeException {
+    public EmptyInputException(String message) {
+        super(message);
+    }
+}
```
Comment (klezner) [inline src/main/java/com/mkgn/weatherservice/exceptions/EmptyInputException.java:7]: Exception is unnecessary

File: src/main/java/com/mkgn/weatherservice/exceptions/CardinalsOutOfRangeException.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.mkgn.weatherservice.exceptions;
+
+public class CardinalsOutOfRangeException extends RuntimeException {
+    public CardinalsOutOfRangeException(String message) {
+        super(message);
+    }
+}
```
Comment (klezner) [inline src/main/java/com/mkgn/weatherservice/exceptions/CardinalsOutOfRangeException.java:7]: Exception is unnecessary

File: src/main/java/com/mkgn/weatherservice/exceptions/EmptyInputException.java (hunk 1 above)
Comment (MichalKulygin) [reply src/main/java/com/mkgn/weatherservice/exceptions/EmptyInputException.java:7]: OK, removed

File: src/main/java/com/mkgn/weatherservice/exceptions/CardinalsOutOfRangeException.java (hunk 2 above)
Comment (MichalKulygin) [reply src/main/java/com/mkgn/weatherservice/exceptions/CardinalsOutOfRangeException.java:7]: OK, removed

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                        .region("Region")
+                        .country("Country")
+                        .latitude(54.0)
+                        .longitude(18.0)
+                        .build());
+        // when
+        locationService.createLocation("City", "Region", "Country", 54.0, 18.0);
+        //then
+        verify(locationRepository).save(locationArgumentCaptor.capture());
+        assertThat(locationArgumentCaptor.getValue().getCity().equals("City"));
```
Comment (klezner) [inline src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:41]: create a variable for locationArgumentCaptor.getValue()

Comment (MichalKulygin) [reply src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:41]: done

**Review by klezner — APPROVED** [review]

**Review by klezner — APPROVED** [review]

File: src/main/java/com/mkgn/weatherservice/location/LocationController.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        String region = request.getRegion();
+        String country = request.getCountry();
+        Double latitude = request.getLatitude();
+        Double longitude = request.getLongitude();
+
+        Location location = locationService.createLocation(city, region, country, latitude, longitude);
+
+        NewLocationResponse responseBody = locationMapper.mapLocationToNewLocationResponse(location);
+
+        return ResponseEntity.status(201).body(responseBody);
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationController.java:31]: 201 -> HttpStatus.CREATED

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+class LocationService {
+
+    private final LocationRepository locationRepository;
+
+    Location createLocation(String city, String region, String country, Double latitude, Double longitude) {
+
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationService.java:13]: unnecessary new line

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        Location location = new Location();
+        location.setCity(city);
+        location.setRegion(region);
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationService.java:16]: it saves trash values like "   "

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    @InjectMocks
+    LocationService locationService;
+    @Captor
+    ArgumentCaptor<Location> locationArgumentCaptor;
+
+    @Test
+    void createLocation_thenCreatesNewLocation() {
+        // given
+        when(locationRepository.save(any())).thenReturn(
+
```
Comment (MisQl) [inline src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:29]: unnecessary new line

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                        .latitude(54.0)
+                        .longitude(18.0)
+                        .build());
+        // when
+        locationService.createLocation("City", "", "Country", 54.0, 18.0);
+        //then
+        verify(locationRepository).save(locationArgumentCaptor.capture());
+        Location location = locationArgumentCaptor.getValue();
+        assertThat(location.getCity().equals("City"));
+        assertThat(location.getRegion().equals(""));
```
Comment (MisQl) [inline src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:67]: it should be a null value

File: src/main/java/com/mkgn/weatherservice/location/LocationController.java (hunk 4 above)
Comment (MichalKulygin) [reply src/main/java/com/mkgn/weatherservice/location/LocationController.java:31]: OK

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java (hunk 8 above)
Comment (MichalKulygin) [reply src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:67]: Renamed test method to createLocation_whenRegionIsEmpty_thenCreatesNewLocation()

**Review by klezner — APPROVED** [review]

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java (hunk 8 above)
Comment (MisQl) [reply src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:65]: it should be a null value

Comment (klezner) [reply src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:65]: location.getRegion() == null

**Review by klezner — CHANGES_REQUESTED** [review]

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
```
Comment (klezner) [inline src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:65]: location.getRegion() == null

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java (hunk 8 above)
Comment (MichalKulygin) [reply src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:65]: fixed

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+                        .latitude(54.0)
+                        .longitude(18.0)
+                        .build());
+        // when
+        locationService.createLocation("City", "", "Country", 54.0, 18.0);
+        //then
+        verify(locationRepository).save(locationArgumentCaptor.capture());
+        Location location = locationArgumentCaptor.getValue();
+        assertThat(location.getCity().equals("City"));
+        assertThat(location.getRegion() == null);
```
Comment (MisQl) [inline src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:65]: assertThat(location.getRegion()).isNull();

Comment (MichalKulygin) [reply src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java]: fixed

**Comment by MisQl** [comment]
is OK

**Review by MisQl — APPROVED** [review]

**Pull request merged**